HeartbeatLogging/Sources/HeartbeatController.swift

```diff
@@ -14,92 +14,104 @@
 
 import Foundation
 
-/// A  logger object that provides API to log and flush heartbeats from a synchronized storage container.
+/// An object that provides API to log and flush heartbeats from a synchronized storage container.
 public final class HeartbeatController {
   /// The thread-safe storage object to log and flush heartbeats from.
   private let storage: HeartbeatStorageProtocol
-  // TODO: Document.
-  private let limit: Int = 30 // TODO: Decide on default value.
-  /// Current date provider. It is used instead of `Date.init` for testability.
+  // TODO: Decide on default value.
+  /// The max capacity of heartbeats to store in storage.
+  private let heartbeatsStorageCapacity: Int = 30
+  /// Current date provider. It is used for testability.
   private let dateProvider: () -> Date
-  // TODO: Verify that this standardization aligns with backend.
+  // TODO: Maybe share config with HeartbeatsPayload's DateFormatter?
   /// Used for standardizing dates for calendar-day comparision.
-  static let dateStandardizer = Calendar(identifier: .gregorian).startOfDay(for:)
+  static let dateStandardizer: (Date) -> (Date) = {
+    var calendar = Calendar(identifier: .iso8601)
+    calendar.locale = Locale(identifier: "en_US_POSIX")
+    calendar.timeZone = TimeZone(secondsFromGMT: 0)!
+    return calendar.startOfDay(for:)
+  }()
 
   /// Public initializer.
-  ///
-  /// - Parameter id: The `id` to associate this logger's internal storage with.
+  /// - Parameter id: The `id` to associate this controller's heartbeat storage with.
   public convenience init(id: String) {
-    // TODO: Sanitize id.
+    self.init(id: id, dateProvider: Date.init)
+  }
+
+  /// Convenience initializer. Mirrors the semantics of the public intializer with the added benefit of
+  /// injecting a custom date provider for improved testability.
+  /// - Parameters:
+  ///   - id: The id to associate this controller's heartbeat storage with.
+  ///   - dateProvider: A date provider.
+  convenience init(id: String,
+                   dateProvider: @escaping () -> Date) {
```

Comment:
> I wonder where this should be used compared to the one below

So for integration tests, my understanding was I should use the public API. I felt I couldnt test important stuff though (like logging heartbeats over a period of time) by using the public initializer alone because it does not allowing injecting a `dateProvider`. 

So to compromise, I created this internal initializer that the **public init calls directly into** and **allows for injecting a `dateProvider`**. 

So, to answer your question, this initializer should be used in integration tests.

> since from a testability standpoint this is accessing a singleton (using shared storage) and could conflict with other tests.

That's correct. To avoid this, I use pass `#function` as the `id` in test methods to ensure they are getting an instance that is specific to the specific test case. 

https://github.com/firebase/firebase-ios-sdk/blob/6f1c4eb1187a9056ac9cc07a4f98102cab1467c2/HeartbeatLogging/Tests/Integration/HeartbeatLoggingIntegrationTests.swift#L68-L71


LMK what you think and I'm very open to feedback on this approach. So far, it did come in handy for writing a test like this:
https://github.com/firebase/firebase-ios-sdk/blob/6f1c4eb1187a9056ac9cc07a4f98102cab1467c2/HeartbeatLogging/Tests/Integration/HeartbeatLoggingIntegrationTests.swift#L131-L206

Reply:
Great, I'll keep it as is. I have a test that does the exact thing you describe to verify this behavior. Injecting the date prevents risk of flake here so we can test the instance management logic.

https://github.com/firebase/firebase-ios-sdk/blob/0c827b651262bb0fafc9ba8c14190659cd7bb9a2/HeartbeatLogging/Tests/Integration/HeartbeatLoggingIntegrationTests.swift#L73-L96